File: packages/core/test/policy.test.ts
import { describe, expect, test } from "bun:test";
import { gateApproval, route, selectPolicy, NoPolicyError } from "../src/policy";
import { DANA, MEL, PRIYA, ROSTER, TIERS, member, policy } from "./fixtures";

const invoice = (amount: string, submitterId = PRIYA.id) => ({
  id: "inv_1", amount, submitterId,
});

describe("selectPolicy", () => {
  test("picks the cheapest tier that still covers the amount", () => {
    expect(selectPolicy(TIERS, 100).name).toBe("Fast lane");
    expect(selectPolicy(TIERS, 500).name).toBe("Fast lane");   // boundary is inclusive
    expect(selectPolicy(TIERS, 501).name).toBe("Standard");
    expect(selectPolicy(TIERS, 5000).name).toBe("Standard");
    expect(selectPolicy(TIERS, 5001).name).toBe("High value");
  });

  test("an amount above every ceiling falls to the strictest tier, never to none", () => {
    expect(selectPolicy(TIERS, 10_000_000).name).toBe("High value");
  });

  test("ignores inactive tiers", () => {
    const tiers = [policy({ id: "p_off", name: "Fast lane", maxAmount: "500", active: false }), TIERS[1]!];
    expect(selectPolicy(tiers, 100).name).toBe("Standard");
  });

  test("throws when the org has no active policy at all", () => {
    expect(() => selectPolicy([], 100)).toThrow(NoPolicyError);
  });
});

describe("route", () => {
  test("the submitter is never among the eligible approvers", () => {
    const d = route({ invoice: invoice("2400", MEL.id), policies: TIERS, members: ROSTER });
    expect(d.eligibleApprovers.map((m) => m.id)).toEqual([DANA.id]);
  });

  test("members without an approver role are excluded", () => {
    const d = route({ invoice: invoice("100"), policies: TIERS, members: ROSTER });
    expect(d.eligibleApprovers.map((m) => m.id).sort()).toEqual([DANA.id, MEL.id].sort());
  });

  test("required approvals never exceed the number of people who can give them", () => {
    const d = route({ invoice: invoice("20000", DANA.id), policies: TIERS, members: ROSTER });
    expect(d.policy.name).toBe("High value");
    expect(d.eligibleApprovers).toHaveLength(1); // only Mel is left
    expect(d.requiredApprovals).toBe(1);         // clamped down from 2, not deadlocked
  });

  test("carries the tier's attestation requirement", () => {
    expect(route({ invoice: invoice("100"), policies: TIERS, members: ROSTER }).requiredAttestation).toBeNull();
    expect(route({ invoice: invoice("2400"), policies: TIERS, members: ROSTER }).requiredAttestation).toBe("selfie_check");
  });

  test("explains itself in one human-readable line", () => {
    const d = route({ invoice: invoice("2400"), policies: TIERS, members: ROSTER });
    expect(d.reason).toContain("Standard");
    expect(d.reason).toContain("selfie check");
  });
});

describe("gateApproval", () => {
  const base = route({ invoice: invoice("2400"), policies: TIERS, members: ROSTER });

  const gate = (over: Partial<Parameters<typeof gateApproval>[0]> = {}) =>
    gateApproval({
      decision: base,
      approver: MEL,
      submitterId: PRIYA.id,
      alreadyDecided: false,
      attestationAgeSec: 10,
      ...over,
    });

  test("allows a fresh, eligible, first-time approval", () => {
    expect(gate().allowed).toBe(true);
  });

  test("blocks self-approval", () => {
    const d = route({ invoice: invoice("2400", MEL.id), policies: TIERS, members: ROSTER });
    expect(gate({ decision: d, submitterId: MEL.id }).code).toBe("self_approval");
  });

  test("blocks someone who isn't an approver on this tier", () => {
    // Not the submitter, so this can only be the role check firing.
    const sam = member({ id: "sam", role: "finance" });
    expect(gate({ approver: sam }).code).toBe("not_eligible");
  });

  test("blocks a second decision from the same approver", () => {
    expect(gate({ alreadyDecided: true }).code).toBe("already_decided");
  });

  test("blocks approval with no selfie check when the tier demands one", () => {
    expect(gate({ attestationAgeSec: null }).code).toBe("attestation_required");
  });

  test("blocks a stale selfie check", () => {
    expect(gate({ attestationAgeSec: 301 }).code).toBe("attestation_stale");
    expect(gate({ attestationAgeSec: 299 }).allowed).toBe(true);
  });

  test("the fast lane needs no attestation at all", () => {
    const d = route({ invoice: invoice("100"), policies: TIERS, members: ROSTER });
    expect(gate({ decision: d, attestationAgeSec: null }).allowed).toBe(true);
  });
});

File: server/internal/policy/policy_test.go
package policy

import (
	"testing"

	"github.com/Nishu0/quorly/server/internal/domain"
)

func tiers() []domain.Policy {
	return DefaultTiers()
}

var (
	dana  = domain.Member{ID: "dana", Role: domain.RoleOwner, Email: "dana@acme.test"}
	mel   = domain.Member{ID: "mel", Role: domain.RoleApprover, Email: "mel@acme.test"}
	priya = domain.Member{ID: "priya", Role: domain.RoleMember, Email: "priya@acme.test"}
	sam   = domain.Member{ID: "sam", Role: domain.RoleFinance, Email: "sam@acme.test"}
)

func roster() []domain.Member { return []domain.Member{dana, mel, priya} }

func invoice(amount float64, submitter string) domain.Invoice {
	return domain.Invoice{ID: "inv_1", Amount: amount, SubmitterID: submitter}
}

func TestSelectPicksCheapestCoveringTier(t *testing.T) {
	cases := []struct {
		amount float64
		want   string
	}{
		{100, "Fast lane"},
		{500, "Fast lane"}, // boundary is inclusive
		{501, "Standard"},
		{5000, "Standard"},
		{5001, "High value"},
	}
	for _, c := range cases {
		got, err := Select(tiers(), c.amount)
		if err != nil {
			t.Fatalf("amount %v: %v", c.amount, err)
		}
		if got.Name != c.want {
			t.Errorf("amount %v: got %q, want %q", c.amount, got.Name, c.want)
		}
	}
}

func TestSelectFallsToStrictestAboveEveryCeiling(t *testing.T) {
	got, err := Select(tiers(), 10_000_000)
	if err != nil {
		t.Fatal(err)
	}
	if got.Name != "High value" {
		t.Errorf("got %q, want the strictest tier", got.Name)
	}
}

func TestSelectIgnoresInactiveTiers(t *testing.T) {
	ts := tiers()
	ts[0].Active = false
	got, err := Select(ts, 100)
	if err != nil {
		t.Fatal(err)
	}
	if got.Name != "Standard" {
		t.Errorf("got %q, want Standard", got.Name)
	}
}

func TestSelectErrorsWithNoActivePolicy(t *testing.T) {
	if _, err := Select(nil, 100); err != ErrNoPolicy {
		t.Errorf("got %v, want ErrNoPolicy", err)
	}
}

func TestRouteExcludesTheSubmitter(t *testing.T) {
	d, err := Route(invoice(2400, mel.ID), tiers(), roster())
	if err != nil {
		t.Fatal(err)
	}
	if len(d.EligibleApprovers) != 1 || d.EligibleApprovers[0].ID != dana.ID {
		t.Errorf("got %+v, want only dana", d.EligibleApprovers)
	}
}

func TestRouteExcludesNonApproverRoles(t *testing.T) {
	d, err := Route(invoice(100, priya.ID), tiers(), roster())
	if err != nil {
		t.Fatal(err)
	}
	if len(d.EligibleApprovers) != 2 {
		t.Errorf("got %d approvers, want dana and mel", len(d.EligibleApprovers))
	}
}

func TestRouteClampsRequiredApprovalsToAvailablePeople(t *testing.T) {
	// High value wants 2, but with Dana submitting only Mel remains.
	d, err := Route(invoice(20000, dana.ID), tiers(), roster())
	if err != nil {
		t.Fatal(err)
	}
	if d.Policy.Name != "High value" {
		t.Fatalf("got tier %q", d.Policy.Name)
	}
	if len(d.EligibleApprovers) != 1 {
		t.Fatalf("got %d eligible", len(d.EligibleApprovers))
	}
	if d.RequiredApprovals != 1 {
		t.Errorf("got %d required, want 1 — otherwise the invoice deadlocks", d.RequiredApprovals)
	}
}

func TestRouteCarriesAttestationRequirement(t *testing.T) {
	fast, _ := Route(invoice(100, priya.ID), tiers(), roster())
	if fast.RequiredAttestation != nil {
		t.Error("fast lane should need no biometric")
	}
	std, _ := Route(invoice(2400, priya.ID), tiers(), roster())
	if std.RequiredAttestation == nil || *std.RequiredAttestation != domain.SelfieCheck {
		t.Error("standard tier should require a selfie check")
	}
}

func TestGateAllowsFreshEligibleFirstDecision(t *testing.T) {
	d, _ := Route(invoice(2400, priya.ID), tiers(), roster())
	age := 10
	got := Gate(GateInput{Decision: d, Approver: mel, SubmitterID: priya.ID, AttestationAgeSec: &age})
	if !got.Allowed {
		t.Errorf("got %+v, want allowed", got)
	}
}

func TestGateBlocksSelfApproval(t *testing.T) {
	d, _ := Route(invoice(2400, mel.ID), tiers(), roster())
	age := 10
	got := Gate(GateInput{Decision: d, Approver: mel, SubmitterID: mel.ID, AttestationAgeSec: &age})
	if got.Code != GateSelfApproval {
		t.Errorf("got %q, want self_approval", got.Code)
	}
}

func TestGateBlocksNonApprover(t *testing.T) {
	d, _ := Route(invoice(2400, priya.ID), tiers(), roster())
	age := 10
	// Sam isn't the submitter, so this can only be the role check firing.
	got := Gate(GateInput{Decision: d, Approver: sam, SubmitterID: priya.ID, AttestationAgeSec: &age})
	if got.Code != GateNotEligible {
		t.Errorf("got %q, want not_eligible", got.Code)
	}
}

func TestGateBlocksSecondDecisionFromSameApprover(t *testing.T) {
	d, _ := Route(invoice(2400, priya.ID), tiers(), roster())
	age := 10
	got := Gate(GateInput{Decision: d, Approver: mel, SubmitterID: priya.ID, AlreadyDecided: true, AttestationAgeSec: &age})
	if got.Code != GateAlreadyDecided {
		t.Errorf("got %q, want already_decided", got.Code)
	}
}

func TestGateRequiresAttestationWhenTierDemandsIt(t *testing.T) {
	d, _ := Route(invoice(2400, priya.ID), tiers(), roster())
	got := Gate(GateInput{Decision: d, Approver: mel, SubmitterID: priya.ID, AttestationAgeSec: nil})
	if got.Code != GateAttestationRequired {
		t.Errorf("got %q, want attestation_required", got.Code)
	}
}

func TestGateRejectsStaleAttestation(t *testing.T) {
	d, _ := Route(invoice(2400, priya.ID), tiers(), roster())
	stale, fresh := 301, 299
	if got := Gate(GateInput{Decision: d, Approver: mel, SubmitterID: priya.ID, AttestationAgeSec: &stale}); got.Code != GateAttestationStale {
		t.Errorf("301s: got %q, want attestation_stale", got.Code)
	}
	if got := Gate(GateInput{Decision: d, Approver: mel, SubmitterID: priya.ID, AttestationAgeSec: &fresh}); !got.Allowed {
		t.Errorf("299s: got %+v, want allowed", got)
	}
}

func TestGateSkipsAttestationOnFastLane(t *testing.T) {
	d, _ := Route(invoice(100, priya.ID), tiers(), roster())
	got := Gate(GateInput{Decision: d, Approver: mel, SubmitterID: priya.ID, AttestationAgeSec: nil})
	if !got.Allowed {
		t.Errorf("got %+v, want allowed without a biometric", got)
	}
}
